src/storage/index/IndexPolicyMaker.cpp

```diff
@@ -41,41 +41,58 @@ cpp2::ErrorCode IndexPolicyMaker::decodeExpression(const std::string &filter) {
     return code;
 }
 
-void IndexPolicyMaker::buildPolicy() {
-    prefix_.reserve(256);
-    decltype(operatorList_.size()) hintNum = 0;
-    bool hasStr = false;
+bool IndexPolicyMaker::buildPolicy() {
+    bool nextCol = true;
     for (auto& col : index_->get_fields()) {
-        auto it = std::find_if(operatorList_.begin(), operatorList_.end(),
-                               [&col] (const auto& tup) {
-                                   return col.get_name() == std::get<0>(tup);;
-                               });
-        if (it != operatorList_.end()) {
-            if (std::get<2>(*it) == RelationalExpression::Operator::EQ) {
+        auto itr = operatorList_.begin();
+        while (nextCol && itr != operatorList_.end()) {
+            if (col.get_name() == std::get<0>(*itr)) {
                 /**
                  * TODO sky : drop the sub-exp from root expression tree.
+                 * TODO (sky) : String range scan was disabled on graph layer.
+                 *              it is not support for storage layer .
                  */
-                hintNum++;
-                auto v = std::get<1>(*it);
-                hasStr = (v.which() == VAR_STR);
-                prefix_.append(NebulaKeyUtils::encodeVariant(v));
+                if (std::get<2>(*itr) == RelationalExpression::Operator::NE) {
+                    requiredFilter_ = true;
+                    nextCol = false;
+                    break;
+                }
+                if (!writeScanItem(col.get_name(), *itr)) {
+                    return false;
+                }
+                // string field still need filter.
+                if (col.get_type().type == nebula::cpp2::SupportedType::STRING) {
+                    requiredFilter_ = true;
+                }
+                // Delete operator item if hint.
+                operatorList_.erase(itr);
             } else {
-                break;
+                ++itr;
             }
-        } else {
+        }
+        /**
+         * If index field does not hit the filter condition,
+         * means there is no need to loop the index fields. for example :
+         * index (c1, c2, c3)
+         * where c1 > 1 and c3 == 1
+         * Field c2 is missing from the operatorList_,
+         * So we just need using c1 to range scan and filter c3.
+         */
+        auto exist = scanItems_.find(col.get_name());
+        if (exist == scanItems_.end()) {
             break;
         }
     }
-    if (optimizedPolicy_ && hintNum == operatorList_.size() && !hasStr) {
-        requiredFilter_ = false;
+    // re-check operatorList_.
+    // if operatorList_ is not empty, that means there are still fields to filter
+    if (!requiredFilter_ && operatorList_.size() > 0) {
+        requiredFilter_ = true;
     }
+    return true;
```

Comment:
need here check wrong filter condition like c1 > 2 and c1 < 1?

Reply:
> need here check wrong filter condition like c1 > 2 and c1 < 1?

I don't think we need to check it here, but if we do, it's better to put it in other method .

Reply:
> need here check wrong filter condition like c1 > 2 and c1 < 1?

I think it's ok even if no error message is returned. It just doesn't return the result set. Current processing  same with some relational databases. WDYT?